### ISSUE TITLE
Query Arg in request URL

### ISSUE THREAD

**Issue by haudiobe** [cta-wave/common-media-client-data#108]
All,

in MPEG-DASH the idea was floated to have the CMCD keys as part of the offered MPD URL for session and content. However, this generally may result in an inconsistency.

Moreover and more general, can a resource that is requested include a CMCD formatted query arg. This may result in conflicts.

It would be good if the specification is more specific about this issue.

**Comment by wilaw** [comment]
To illustrate this problem further, what if the DASH mpd URL, or a segment URL given to the player looks like this

`https://example.com/manifest.mpd?CMCD=sid%3D%2221cf726cfe3d937b5f974f72bb5bd06a%22`

`https://example.com/segments/seg34.mp4?CMCD=sid%3D%2221cf726cfe3d937b5f974f72bb5bd06a%22`

The spec should clarify the following points:
1. The player must not add a new CMCD query arg if one already exists. It should urldecode the existing CMCD arguments, add its own and then urlencode the result.
2. What to do about conflicts? Is a sid is already supplied, should the player preserve it or replace it with a new random  sid?

**Comment by ZmGorynych** [comment]
1. We may want to introduce a DASH query param argument namespace to avoid a collision above. E.g., the query parameter for sid would be `cmcd_sid`. This makes it less likely that we'll get a namespace collision.
```https://example.com/segments/seg34.mp4?cmcd_sid=42```
2. If someone really wants to put a CMCD string via a 302 redirect we cannot prohibit this. Maybe there is a valid deployment reason for this (e.g., reporting only sid and cid). 
3. If we have more than one CMCD query parameter, we can require coalescing the keys together and having client-derived keys have precedence. With that said, this becomes too complex and prone to weird errors, I would rather require outright replacement of "inherited" CMCD query parameter with the one computed by the client. I think it is safer. sid and cid can be specified via special query parameters (e.g., `cmcd_sid=42` above).

The reason for the whole exercise is to avoid random sid and cid values and have them assigned by the provider. This allows correlation between sessions from the same linear channel (e.g. cid is FOX_NEWS or CNN), correlation with internal data (e.g., state of the DOCSIS network in the viewer's neighborhood, as reported by the network equipment), etc. Things can be later easily anonymized, but opaque strings are not really useful.

5. . What was proposed was to have specially named variables

**Comment by gwendalsimon** [comment]
Indeed, is there any key (other than sid and cid) that the server would like to pass to the player? Note that we already have CMSD to let the server send metadata to the player.
The "cmcd_sid" key (instead of "sid") in the url is a good idea imho.

**Comment by PCaponetti** [comment]
Paul C: do we need to insert ourselves in this conversation? If the origin or middle proxy adds these keys in a manifest, the player can receive them and decide whether or not to use them, no?
Nicolas L: this mechanism could be useful for interstitials for interstitials to get content ID
Casey: any manifest-provided CMCD information is likely overwritten by player implementations anyway. This means we should have a stance on what to do in the spec.

Players don't necessarily (likely won't) have decoding code, so merging CMCD from manifest and proxy/origin may not be possible. Maybe origin/proxy supplied CMCD information would only be used as a fallback and work only in players that do not support CMCD.

We'll have to talk more about this. The override/merge rules may be something we need to specify to help player implementers.

Added this to the spec in suggest mode for now: 

```
If there are CMCD keys specified in a manifest as part of segment request URLs care should be taken to remove/overwrite, or merge that with future CMCD reports
```

**Comment by PCaponetti** [comment]
Talked about this and have the following concerns:

If we don't require or urge player implementations to have per key configuration on behavior to use when the upstream defines a CMCD key (keep or overwrite) we put an undue burden on the analysis of this data. There are likely platforms that use multiple players and will expect consistent behavior amongst them. Especially given that there is no way (with CMCD alone) for us to know what player is reporting the CMCD data.

BUT if we urge or require player implementations to have these configuration options, that may be an undue burden for implementers. Also, we have this problem in CMCD v1 and haven't heard of any major issues resulting from us not specifying that this configuration option is necessary.

**Comment by wilaw** [comment]
Meeting: the entire CMCD query arg is overwritten in most players. We do not see a problem here.
